src/Features/Core/Portable/Diagnostics/BuildOnlyDiagnosticsService.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    [ExportWorkspaceServiceFactory(typeof(IBuildOnlyDiagnosticsService), ServiceLayer.Default), Shared]
+    internal sealed class BuildOnlyDiagnosticsServiceFactory : IWorkspaceServiceFactory
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public BuildOnlyDiagnosticsServiceFactory()
+        {
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+            => new BuildOnlyDiagnosticsService(workspaceServices.Workspace);
+
+        private sealed class BuildOnlyDiagnosticsService : IBuildOnlyDiagnosticsService
+        {
+            private readonly object _gate = new();
+            private readonly Dictionary<DocumentId, ImmutableArray<DiagnosticData>> _documentDiagnostics = new();
+            private readonly Dictionary<ProjectId, ImmutableArray<DiagnosticData>> _projectDiagnostics = new();
+
+            public BuildOnlyDiagnosticsService(Workspace workspace)
+            {
+                workspace.WorkspaceChanged += OnWorkspaceChanged;
+            }
+
+            private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+            {
+                switch (e.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                        ClearAllDiagnostics();
+                        break;
+
+                    case WorkspaceChangeKind.ProjectReloaded:
+                    case WorkspaceChangeKind.ProjectRemoved:
+                        ClearDiagnostics(e.OldSolution.GetProject(e.ProjectId));
+                        break;
+
+                    case WorkspaceChangeKind.DocumentRemoved:
+                    case WorkspaceChangeKind.DocumentReloaded:
+                    case WorkspaceChangeKind.AdditionalDocumentRemoved:
+                    case WorkspaceChangeKind.AdditionalDocumentReloaded:
+                    case WorkspaceChangeKind.AnalyzerConfigDocumentRemoved:
+                    case WorkspaceChangeKind.AnalyzerConfigDocumentReloaded:
```

Comment:
I am not familiar with how this works but is it ok to only update current document in such case? Adding or removing an additional document could have a project level impact, isn't it?

Reply:
Good point, let me ensure we clear entire project diagnostics for this case.